php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+                    } else {
+                        $skillCount++;
+                    }
+                }
+
+                $skillCount = min($Skill->getDemonstrationsRequiredByLevel($this->Level), $skillCount);
+            }
+
+            $count += $skillCount;
+        }
+
+        return $this->demonstrationsComplete = $count;
+    }
+
+    private $demonstrationsAverage;
+    public function getDemonstrationsAverage()
+    {
+        if (!empty($this->demonstrationsAverage)) {
+            return $this->demonstrationsAverage;
+        }
+
+        $effectiveDemonstrationsData = $this->getEffectiveDemonstrationsData();
+        $average = null;
+        $totalScore = 0;
+
+        foreach ($effectiveDemonstrationsData as $skillId => $demonstrationsData) {
+            foreach ($demonstrationsData as $demonstration) {
+                if (empty($demonstration['Override'])) {
+                    $totalScore += $demonstration['DemonstratedLevel'];
+                }
+            }
+        }
+
+        if ($this->getDemonstrationsLogged()) {
+            $average = $totalScore / $this->getDemonstrationsLogged();
+        }
+
+        return $this->demonstrationsAverage = $average;
+    }
+
+    public function isLevelComplete()
+    {
+        $logged = $this->getDemonstrationsLogged();
+        $completed = $this->getDemonstrationsComplete();
+        $average = $this->getDemonstrationsAverage();
+
+        $competencyEvidenceRequirements = $this->Competency->getTotalDemonstrationsRequired($this->Level);
+        $minimumOffset = $this->Competency->getMinimumAverageOffset();
 
         return (
-                $completion['demonstrationsComplete'] >= $Competency->getTotalDemonstrationsRequired($completion['currentLevel']) &&
-                (
-                    $completion['demonstrationsLogged'] == 0 || // if demonstrationsComplete is full but none are logged, the student has fulfilled all their demonstrations via overrides and the average is irrelevant
-                    $completion['demonstrationsAverage'] >= ($completion['currentLevel'] + $Competency->getMinimumAverageOffset())
-                )
+            $completed >= $comptencyEvidenceRequirements &&
+            (
+                $logged === 0 ||
+                $average >= ($this->Level + $minimumOffset)
+            )
         );
     }
+
+    public function getGrowth()
+    {
+        if (empty($this->BaselineRating)) {
+            return null;
+        }
+
+        $demonstrationData = $this->getDemonstrationData();
+
+        $totalSkills = $this->Competency->getTotalSkills();
+
+        $growthData = array_map(function($demonstrations, $skillId) {
+            if (count($demonstrations) === 1 && empty($this->BaselineRating)) {
+                return null;
+            }
+
+            $lastRating = end($demonstrations);
+            if ($this->BaselineRating) {
+                return $lastRating['DemonstratedLevel'] - floatval($this->BaselineRating);
```

Comment:
`floatval` won't be needed here after the model field is fixed to return correct type